### PR TITLE
implementation of the onStartLoad handler for gatsby-image

### PR DESCRIPTION
**Pull request by kaoDev**
quick implementation for the feature request from the issue https://github.com/gatsbyjs/gatsby/issues/6659. 
I think this should be enough to handle the scenario described by @zionis137

With this change the `gatsby-image` component accepts a new optional prop `onStartLoad`, it is a function called with no arguments and expecting no return value.

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Deploy preview for *using-postcss-sass* failed.

Built with commit 596d40623f6a6c33ff635314270e32c835d0f039

https://app.netlify.com/sites/using-postcss-sass/deploys/5b5776d84ed62f14b743e686

**Comment by KyleAMathews** [comment]
Deploy preview for *gatsbyjs* failed.

Built with commit 596d40623f6a6c33ff635314270e32c835d0f039

https://app.netlify.com/sites/gatsbyjs/deploys/5b5776d74ed62f14b743e67e

File: packages/gatsby-image/src/index.js [hunk 1]
```diff
@@ -168,9 +168,19 @@ class Image extends React.Component {
     this.handleRef = this.handleRef.bind(this)
   }
 
+  componentDidMount() {
+    if (this.state.isVisible && typeof this.props.onStartLoad === `function`) {
+      this.props.onStartLoad()
```
Comment (pieh) [inline packages/gatsby-image/src/index.js]: just checking here - this won't cause firing this "event" twice if image is visible in initial viewarea?

**Comment by pieh** [comment]
Should it fire if image is already cached in browser?

File: packages/gatsby-image/src/index.js (hunk 1 above)
Comment (kaoDev) [reply packages/gatsby-image/src/index.js]: good point, I just pushed a change to fire only if the image was not visible before

**Comment by zionis137** [comment]
Or the function could have an boolean argument `wasCached`. Then the user can decide what he wants to do.

**Comment by kaoDev** [comment]
the onStartLoad handler now gets called with an object holding the wasCached information, by implementing this I also realised that the information if the image was already seen was written at the wrong moment. Instead of updating the cache state in the onLoad handler the cache entry was set in the constructor, whenever another gatsby-image with the same src was already constructed.

**Comment by kaoDev** [comment]
those errors seem to have the origin in some other unrelated code, I tried to merge a clean master state (a release tag), but there are still errors.
So @KyleAMathews can you please have a look on this or give a hint how this automatic tests/deployments can be fixed on this branch?

**Comment by kaoDev** [comment]
hey @pieh can you have another look, I just cleaned up the rebase mess I pushed a few days back and all checks are green 😄

**Pull request merged**